### PR TITLE
Adding google-competition solution file

### PR DESCRIPTION
**Pull request by paulmartin4**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+from copy import deepcopy
+
+def primary_forbidden_sequences(list_of_all_forbidden_sequences,P):
```
Comment (louismartin) [inline main.py:3]: NIT:
`list_of_all_forbidden_sequences` -> `forbidden_sequences` (naming it as a plural is often clear enough)?
Don't forget the space after the coma :p

File: main.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    all_sequences = deepcopy(list_of_all_forbidden_sequences)
```
Comment (louismartin) [inline main.py:4]: You could actually keep the same name here

File: main.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+def total_number_of_possible_sequence(sequences):
+    impossible_number = 0
+    for sequence in sequences:
+        impossible_number += number_of_impossible_sequences(sequence)
+    return (2**N - impossible_number)
+
+t = int(input())
+
+for i in range(1,t+1):
```
Comment (louismartin) [inline main.py:48]: often better to use the vanilla `for i in range(t):` and use `i + 1` afterwards

File: main.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    N,P = [int(s) for s in input().split(" ")]
```
Comment (louismartin) [inline main.py:49]: I know you don't like it but forgetting spaces after comas is really a signal of bad coding...

File: main.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    list_of_all_forbidden_sequences = []
+    for j in range(P):
+        list_of_all_forbidden_sequences.append(list(input()))
+    print("Case #{}: {}".format(i,total_number_of_possible_sequence(primary_forbidden_sequences(list_of_all_forbidden_sequences,P))))
```
Comment (louismartin) [inline main.py:53]: I would add intermediary lines for the code to be more readable.

File: main.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+from copy import deepcopy
+
+def primary_forbidden_sequences(list_of_all_forbidden_sequences,P):
+    all_sequences = deepcopy(list_of_all_forbidden_sequences)
+    index_to_be_taken = [0]*P
+    for i in range(P):
+        if index_to_be_taken[i] == 0:
```
Comment (louismartin) [inline main.py:7]: Won't this test always be true?

File: main.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            current_sequence = all_sequences[i]
```
Comment (louismartin) [inline main.py:8]: It is more natural to write `for sequence in all_sequences:`

File: main.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+            n = len(current_sequence)
+            working_sequences = deepcopy(all_sequences)
```
Comment (louismartin) [inline main.py:10]: Why do you use deepcopies?

File: main.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            for j in range(n):
```
Comment (louismartin) [inline main.py:11]: As `n` is only used here I think, it might be simpler just to write `for j in range(len(current_sequence)):`

File: main.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+                if len(working_sequences) > 1:
+                    working_sequences = [sequence for sequence in working_sequences if (len(sequence) >= j+1)] # We keep only the long enough sequences
+                    """bis = []
+                    for k in range(len(working_sequences)):
+                        if len(working_sequences[k]) >= j+1:
+                            bis.append(working_sequences[k])
+                    working_sequences = bis"""
+                    working_sequences = [sequence for sequence in working_sequences if current_sequence[j] == sequence [j]] #We keep only the sequences that match the j first characters
```
Comment (louismartin) [inline main.py:19]: You can put the comment on the line before, more readable

**Review by louismartin — APPROVED** [review]
Nice!
I mainly put comments on superficial aspects of the code because I didn't dive into the logic, seems good anyway.
I don't know if it is possible for this specific problem but you should try to use less nested loops, they make the code way harder to read.
One solution is to use more small functions:
e.g. instead of writing something like:
```
results = []
for i in range(n):
    nested_result = 0
    for k in range(m):
        nested_result += ...
    results.append(nested_result)
```
Write
```
def explicit_method_for_nested_result():
    result = 0
    for j in range(m):
        result += ...
    return result


result = []
for i in range(n):
    result.append(explicit_method_for_nested_result())
```
and even better with list comprehensions when your code is modular and clean enough:
```
def explicit_method_for_nested_result():
    return sum([... for j in range(m)])


result = [explicit_method_for_nested_result() for i in range(n)]
```

File: main.py (hunk 7 above)
Comment (paulmartin4) [reply main.py:8]: If I do write `for sequence in all_sequences:` then I won't be able to access to my `index_to_be_taken[i]` would I ?

File: main.py (hunk 8 above)
Comment (paulmartin4) [reply main.py:10]: It's just the way I learned to do in prépa when I wanted to be sure my initial list wouldn't change. Would copies be enough ?

**Pull request merged**